### PR TITLE
[TUTORIAL] Gluon performance tips and tricks

### PR DESCRIPTION
**Pull request by thomelane**
Summary of performance features provided in MXNet that are relevant for Gluon users. Useful for a beginner audience who might not know what features to search for in the first place.

## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by thomelane** [comment]
@aaronmarkham would be great if you could review this new tutorial too, cheers! i'll make adjustments to my sparse tutorial tomorrow.

**Comment by pengzhao-intel** [comment]
It's great to have these documents. I think we still need to cover CPU performance.
@xinyu-intel will help to provide the additional information about CPU

**Comment by thomelane** [comment]
Would be great to get CPU specific tricks and tips @pengzhao-intel. Given the length of this tutorial already let's add those in another tutorial as a follow up. @xinyu-intel are you able to write about some performance tricks for CPU and follow up in another PR? When we have that we can think about splitting this up into generic tips, GPU specific and CPU specific tricks.

**Comment by pengzhao-intel** [comment]
> Would be great to get CPU specific tricks and tips @pengzhao-intel. Given the length of this tutorial already let's add those in another tutorial as a follow up. @xinyu-intel are you able to write about some performance tricks for CPU and follow up in another PR? When we have that we can think about splitting this up into generic tips, GPU specific and CPU specific tricks.

Agree. I see lots of tips and tricks are common for both CPU and GPU; so we can make the user aware of this :)

File: docs/tutorials/gluon/performance.md [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+    
+    average iterations/sec: 6.5968
+    average samples/sec: 422.1925
+
+
+Using the above tips and tricks we managed to increase the throughput of training by ~500% from the initial benchmark! Our training throughput is less than the throughput of the individual components we tested, but there are additional overheads that we didn't previously measure (such as data transfer to GPU).
+
+## Conclusion
+
+We learnt a number of tips and tricks to optimize the throughput of training, and they lead to a considerable increase compared to our initial baseline. As general rules, set `num_workers` on the `DataLoader` to >0, and hybridize your network if you're not debugging. You should increase `batch_size` where possible, but do this with care because of its potential effects on convergence. And finally, consider mixed precision training for substantial speed-ups if you're willing to accept a small drop in network accuracy.
```
Comment (aaronmarkham) [inline docs/tutorials/gluon/performance.md]:
```suggestion
We learned a number of tips and tricks to optimize the throughput of training, and they lead to a considerable increase compared to our initial baseline. As general rules, set `num_workers` on the `DataLoader` to >0, and hybridize your network if you're not debugging. You should increase `batch_size` where possible, but do this with care because of its potential effects on convergence. And finally, consider mixed precision training for substantial speed-ups if you're willing to accept a small drop in network accuracy.
```

**Comment by aaronmarkham** [comment]
I restarted the failed CI test for you.

File: docs/tutorials/gluon/performance.md (hunk 1 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Changed.

File: docs/tutorials/gluon/performance.md [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            break
+        print('.', end='')
+    end_time = time.time()
+    total_time = end_time - start_time
+    print('\n')
+    print('average iterations/sec: {:.4f}'.format(num_iters/total_time))
+    print('average samples/sec: {:.4f}'.format(num_samples/total_time))
+```
+
+Our `benchmark_dataloader` function just loops through the `DataLoader` for a given number of iterations: it doesn't transfer the data to the correct context or pass it to the network. Our `benchmark_network` function just performs a forward and backward pass on an identical (and pre-transfered) batch of data: it doesn't require new data to be loaded. We'll run both of these functions now.
```
[the remark this answers is not in the document]
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Changed.

File: docs/tutorials/gluon/performance.md [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+We'll start by writing some code to train an image classification network for the CIFAR-10 dataset, and then benchmark the throughput of the network in terms of samples processed per second. After some performance analysis, we'll identify the bottlenecks (i.e. the components limiting throughput) and improve the training speed step-by-step. We'll bring together all the tips and tricks at the end and evaluate our performance gains.
+
+
+```python
+from __future__ import print_function
+import multiprocessing
+import time
+import mxnet as mx
+import numpy as np
+from PIL import Image
```
[the remark this answers is not in the document]
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Great catch! Added that to perform a slow rotation augmentation (that isn't an MXNet transform), but changed to a `sleep` instead. Switched out for MXNet function, removed dependency, and re-ran the notebook.

File: docs/tutorials/gluon/performance.md [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+
+```python
+from __future__ import print_function
+import multiprocessing
+import time
+import mxnet as mx
+import numpy as np
+```
+
+An Amazon EC2 p3.2xlarge instance was used to benchmark the code in this tutorial. You are likely to get difference results and find different bottlenecks on other hardware, but these tips and tricks should still help improve training speed for bottleneck components. A GPU is recommended for this example.
```
Comment (ptrendx) [inline docs/tutorials/gluon/performance.md]:
```suggestion
An Amazon EC2 p3.2xlarge instance was used to benchmark the code in this tutorial. You are likely to get different results and find different bottlenecks on other hardware, but these tips and tricks should still help improve training speed for bottleneck components. A GPU is recommended for this example.
```

File: docs/tutorials/gluon/performance.md [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+
+```python
+dataset = mx.gluon.data.vision.CIFAR10(train=True)
+print('{} samples'.format(len(dataset)))
+```
+
+    50000 samples
+
+
+So we can learn how to identify training bottlenecks, let's intentionally introduce a short `sleep` into the data loading pipeline. We transform each 32x32 CIFAR-10 image to 244x244 so we can use it with the ResNet-50 network designed for ImageNet. [CIFAR-10 specific ResNet networks](https://gluon-cv.mxnet.io/api/model_zoo.html#gluoncv.model_zoo.get_cifar_resnet) exist but we use the more standard ImageNet variants in this example.
```
Comment (ptrendx) [inline docs/tutorials/gluon/performance.md]: The original ImageNet ResNet actually work on 224x224 images, not 244x244.

File: docs/tutorials/gluon/performance.md [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+    
+     ### benchmark_network 
+    
+    .........
+    
+    average iterations/sec: 13.2545
+    average samples/sec: 212.0723
+
+
+Our data loading pipeline is no longer the bottleneck for training throughput: ~700 samples per second versus ~200 samples per second for network execution as before. We can now focus our attention on improving the network throughput.
```
Comment (ptrendx) [inline docs/tutorials/gluon/performance.md]: Maybe you could add a mention here about [NVIDIA DALI](https://github.com/NVIDIA/DALI), which has MXNet integration and could be useful especially for multi GPU systems?

File: docs/tutorials/gluon/performance.md [hunk 7]
```diff
@@ -0,0 +1,483 @@
...
+
+Overall we see a substantial increase in training throughput: ~85% higher than full-precision training.
+
+## Tips & Tricks #5: Others
+
+Many other tips and tricks exist for optimizing the throughput of training.
+
+One area we didn't explicitly benchmark in this tutorial is data transfer from CPU to GPU memory. Usually this isn't an issue, but for very large arrays this can become a bottleneck too. You might be able to compress your data significantly before transferring if your data is sparse (i.e. mostly zero values). Check out the sparse array tutorial for more details and an example of how this can impact training speed.
+
+Another useful trick if data pre-processing or data transfer is the bottleneck is pre-fetching batches. You can write your training loop to transfer the next batch of data to GPU before processing the current batch. Once again, this trick is memory permitting.
```
Comment (ptrendx) [inline docs/tutorials/gluon/performance.md]: This would actually be very nice to show in this tutorial, since when using the CPU loader it is pretty much free gain (the only resource used is the CPU memory).

File: docs/tutorials/gluon/performance.md [hunk 8]
```diff
@@ -0,0 +1,483 @@
...
+    
+    average iterations/sec: 6.5281
+    average samples/sec: 417.7994
+
+
+Using the above tips and tricks we managed to increase the throughput of training by ~500% from the initial benchmark! Our training throughput is less than the throughput of the individual components we tested, but there are additional overheads that we didn't previously measure (such as data transfer to GPU).
+
+## Conclusion
+
+We learned a number of tips and tricks to optimize the throughput of training, and they lead to a considerable increase compared to our initial baseline. As general rules, set `num_workers` on the `DataLoader` to >0, and hybridize your network if you're not debugging. You should increase `batch_size` where possible, but do this with care because of its potential effects on convergence. And finally, consider mixed precision training for substantial speed-ups if you're willing to accept a small drop in network accuracy.
```
Comment (ptrendx) [inline docs/tutorials/gluon/performance.md]: `if you're willing to accept a small drop in network accuracy` - this is not really true, typically the accuracy of network trained in fp16 and fp32 is basically the same. Do you have any examples of such drops?

File: docs/tutorials/gluon/performance.md (hunk 4 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Good catch.

File: docs/tutorials/gluon/performance.md (hunk 5 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Good catch. Changed, rerun notebook and updated stats.

File: docs/tutorials/gluon/performance.md (hunk 6 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Added.

File: docs/tutorials/gluon/performance.md (hunk 8 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Removed.

File: docs/tutorials/gluon/performance.md (hunk 7 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Quite a complex trick for a beginner tutorial, will leave out details for now. Can follow up with a more advanced performance post later on, since there are many other tricks not discussed here.

**Comment by thomelane** [comment]
thanks for all the reviews @aaronmarkham @pengzhao-intel @ptrendx. updates have been made.

File: tests/tutorials/test_tutorials.py [hunk 9]
```diff
@@ -114,6 +114,9 @@ def test_gluon_save_load_params():
 
 def test_gluon_hybrid():
     assert _test_tutorial_nb('gluon/hybrid')
+
+def test_gluon_hybrid():
```
Comment (wkcn) [inline tests/tutorials/test_tutorials.py]: Thanks for your contribution! It seems to be test_gluon_performance.

File: docs/tutorials/index.md [hunk 10]
```diff
@@ -91,6 +91,7 @@ Select API:&nbsp;
    * [Image similiarity search with InfoGAN](/tutorials/gluon/info_gan.html)
 * Practitioner Guides
     * [Gotchas using NumPy](/tutorials/gluon/gotchas_numpy_in_mxnet.html)
+    * [Performance Tips & Tricks](/tutorials/gluon/performance.html)
```
Comment (pengzhao-intel) [inline docs/tutorials/index.md]: Better to align with the name in the page "Gluon Performance Tips & Tricks"

File: docs/tutorials/gluon/performance.md [hunk 11]
```diff
@@ -0,0 +1,483 @@
...
+```python
+from __future__ import print_function
+import multiprocessing
+import time
+import mxnet as mx
+import numpy as np
+```
+
+An Amazon EC2 p3.2xlarge instance was used to benchmark the code in this tutorial. You are likely to get different results and find different bottlenecks on other hardware, but these tips and tricks should still help improve training speed for bottleneck components. A GPU is recommended for this example.
+
```
Comment (pengzhao-intel) [inline docs/tutorials/gluon/performance.md]: I think both CPU and GPU is fine for this small case.

File: docs/tutorials/gluon/performance.md [hunk 12]
```diff
@@ -0,0 +1,483 @@
...
+<!--- Unless required by applicable law or agreed to in writing, -->
+<!--- software distributed under the License is distributed on an -->
+<!--- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY -->
+<!--- KIND, either express or implied.  See the License for the -->
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
+# Gluon Performance Tips & Tricks
+
+Compared to traditional machine learning methods, the field of deep-learning has increased model accuracy across a wide range of tasks, but it has also increased the amount of computation required for model training and inference. Specialised hardware chips, such as GPUs and FPGAs, can speed up the execution of networks, but it can sometimes be hard to write code that uses the hardware to its full potential. We will be looking at a few simple tips and trick in this tutorial that you can use to speed up training and ultimately save on training costs.
```
Comment (pengzhao-intel) [inline docs/tutorials/gluon/performance.md]: " but it can sometimes be hard to write code that uses the hardware to its full potential."
The framework is the bridge between HW and SW so it supposes it will be easy to write the code. The performance is another thing.

"We will be looking at a few simple tips and trick in this tutorial that you can use to speed up training and ultimately save on training costs."
Does this document only cover for "training"? If so, how about change tilte to "Gluon Performance Tips ... for training"?

File: docs/tutorials/gluon/performance.md [hunk 13]
```diff
@@ -0,0 +1,483 @@
...
+
+```python
+dataset = mx.gluon.data.vision.CIFAR10(train=True)
+print('{} samples'.format(len(dataset)))
+```
+
+    50000 samples
+
+
+So we can learn how to identify training bottlenecks, let's intentionally introduce a short `sleep` into the data loading pipeline. We transform each 32x32 CIFAR-10 image to 224x224 so we can use it with the ResNet-50 network designed for ImageNet. [CIFAR-10 specific ResNet networks](https://gluon-cv.mxnet.io/api/model_zoo.html#gluoncv.model_zoo.get_cifar_resnet) exist but we use the more standard ImageNet variants in this example.
```
Comment (pengzhao-intel) [inline docs/tutorials/gluon/performance.md]: Why add a "sleep"?

File: tests/tutorials/test_tutorials.py (hunk 9 above)
Comment (thomelane) [reply tests/tutorials/test_tutorials.py]: Great catch @wkcn!

Comment (thomelane) [reply tests/tutorials/test_tutorials.py]: Updated.

File: docs/tutorials/index.md (hunk 10 above)
Comment (thomelane) [reply docs/tutorials/index.md]: Updated, thanks.

Comment (thomelane) [reply docs/tutorials/index.md]: Although we're in the Gluon section so adds a bit of noise.

File: docs/tutorials/gluon/performance.md (hunk 11 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Changed to 'tutorial' to be more clear on where the GPU is recommended. Many references to GPU throughout so it's recommended to be able to follow along.

File: docs/tutorials/gluon/performance.md (hunk 12 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: Can use most of these tricks for inference too, added a comment for this.

File: docs/tutorials/gluon/performance.md (hunk 13 above)
Comment (thomelane) [reply docs/tutorials/gluon/performance.md]: It was added to make the data pipeline a bottleneck. Added another comment to clarify.

**Comment by thomelane** [comment]
@pengzhao-intel thanks for the follow-up. made the changes.

**Comment by braindotai** [comment]
Marvellous work by @thomelane, hands down.
Thanks a gazillion times.

**Comment by braindotai** [comment]
A good tutorial on how to convert a Gluon model to Module and vice-versa would be great. I've seen many issues and confusions on Github and MXNet-Forum regarding it. Thanks
